src/api-service/__app__/onefuzzlib/azure/nsg.py

```diff
@@ -0,0 +1,484 @@
+            errors=["cannot update nsg rules. nsg %s not found" % name],
+        )
+
+    logging.info(
+        "setting allowed incoming connection sources for nsg: %s %s",
+        resource_group,
+        name,
+    )
+    all_sources = sources.allowed_ips + sources.allowed_service_tags
+    security_rules = []
+    # NSG security rule priority range defined here:
+    # https://docs.microsoft.com/en-us/azure/virtual-network/network-security-groups-overview
+    min_priority = 100
+    # NSG rules per NSG limits:
+    # https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits?toc=/azure/virtual-network/toc.json#networking-limits
+    max_rule_count = 1000
+    if len(all_sources) > max_rule_count:
+        return Error(
+            code=ErrorCode.INVALID_REQUEST,
+            errors=[
+                "too many rules provided %d. Max allowed: %d"
+                % ((len(all_sources)), max_rule_count),
+            ],
+        )
+
+    priority = min_priority
+    for src in all_sources:
+        security_rules.append(
+            SecurityRule(
+                name="Allow" + str(priority),
+                protocol="*",
+                source_port_range="*",
+                destination_port_range="*",
+                source_address_prefix=src,
+                destination_address_prefix="*",
+                access=SecurityRuleAccess.ALLOW,
+                priority=priority,  # between 100 and 4096
+                direction="Inbound",
+            )
+        )
+        # Will not exceed `max_rule_count` or max NSG priority (4096)
+        # due to earlier check of `len(all_sources)`.
+        priority += 1
+
+    nsg.security_rules = security_rules
+    return update_nsg(nsg)
+
+
+def clear_all_rules(name: str) -> Union[None, Error]:
+    return set_allowed(name, NetworkSecurityGroupConfig())
+
+
+def get_all_rules(name: str) -> Union[Error, List[SecurityRule]]:
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot get nsg rules. nsg %s not found" % name],
+        )
+
+    return cast(List[SecurityRule], nsg.security_rules)
+
+
+def associate_nic(name: str, nic: NetworkInterface) -> Union[None, Error]:
+    resource_group = get_base_resource_group()
+    nsg = get_nsg(name)
+    if not nsg:
+        return Error(
+            code=ErrorCode.UNABLE_TO_FIND,
+            errors=["cannot associate nic. nsg %s not found" % name],
+        )
+
+    if nsg.location != nic.location:
+        return Error(
+            code=ErrorCode.UNABLE_TO_UPDATE,
+            errors=[
+                "network interface and nsg have to be in the same region.",
+                "nsg %s %s, nic: %s %s"
+                % (nsg.name, nsg.location, nic.name, nic.location),
+            ],
+        )
+
+    if nic.network_security_group and nic.network_security_group.id == nsg.id:
+        return None
```

Comment:
How about a log message at some level, like:
```suggestion
        logging.info("NIC %s and NSG %s already associated, not updating", nic.name, name)
        return None
```
This also acts as documentation of what is going on.